pages/docs/guides/working-with-loops.mdx

```diff
@@ -0,0 +1,150 @@
+export const description = 'Each step is executed as a separate HTTP request.';
+
+# Working with Loops in Inngest
+
+In Inngest each step in your function is executed as a separate HTTP request.  This means that for every step in your function, the function is re-entered, starting from the beginning, up to the point where the next step is executed. This [execution model](/docs/learn/how-functions-are-executed) helps in managing retries, timeouts, and ensures robustness in distributed systems.
+
+This page covers how to implement loops in your Inngest functions and avoid common pitfalls.
+
+## Simple function example
+
+Let's start with a simple example to illustrate the concept:
+
+```javascript
+inngest.createFunction(
+  { id: "simple-function" },
+  { event: "test/simple.function" },
+  async ({ step }) => {
+    console.log("hello");
+
+    await step.run("a", async () => { console.log("a") });
+    await step.run("b", async () => { console.log("b") });
+    await step.run("c", async () => { console.log("c") });
+  }
+);
+```
+
+In the above example, you will see "hello" printed four times, once for the initial function entry and once for each step execution (`a`, `b`, and `c`).
+
+
+<CodeGroup forceTabs>
+```bash {{ title: "✅ How Inngest executes the code" }}
+
+# This is how Inngest executes the code above:
+
+<run start>
+"hello"
+
+"hello"
+"a"
+
+"hello"
+"b"
+
+"hello"
+"c"
+<run complete>
+```
+```bash {{ title: "❌ Common incorrect misconception" }}
+
+# This is a common assumption of how Inngest executes the code above.
+# It is not correct.
+
+<run start>
+
+"hello"
+"a"
+"b"
+"c"
+
+<run complete>
+```
+</CodeGroup>
+
+Any non-deterministic logic (like database calls or API calls) must be placed inside a `step.run` call to ensure it is executed correctly within each step.
+
+With this in mind, here is how the previous example can be fixed:
+
+```ts
+inngest.createFunction(
+  { id: "simple-function" },
+  { event: "test/simple.function" },
+  async ({ step }) => {
+    await step.run("hello", () => { console.log("hello") });
+
+    await step.run("a", async () => { console.log("a") });
+    await step.run("b", async () => { console.log("b") });
+    await step.run("c", async () => { console.log("c") });
+  }
+);
+
+// hello
+// a
+// b
+// c
+```
+
+Now, "hello" is printed only once, as expected.
```

Comment:
**suggestion**: Maybe we should include a mention of the logging docs here? We don't recommend console.log, so it could be a good point in time for that. The logger package is designed to prevent duplicate logs.